modal-go/task_command_router_client.go

```diff
@@ -0,0 +1,627 @@
+		stub:         pb.NewTaskCommandRouterClient(conn),
+		conn:         conn,
+		serverClient: serverClient,
+		taskID:       taskID,
+		serverURL:    resp.GetUrl(),
+		logger:       logger,
+	}
+	jwt := resp.GetJwt()
+	client.jwt.Store(&jwt)
+	jwtExp := parseJwtExpiration(ctx, jwt, logger)
+	client.jwtExp.Store(jwtExp)
+
+	logger.DebugContext(ctx, "Successfully initialized command router client", "task_id", taskID)
+	return client, nil
+}
+
+// Close closes the gRPC connection.
+func (c *TaskCommandRouterClient) Close() error {
+	if !c.closed.CompareAndSwap(false, true) {
+		return nil
+	}
+	if c.conn != nil {
+		return c.conn.Close()
+	}
+	return nil
+}
+
+func (c *TaskCommandRouterClient) authContext(ctx context.Context) context.Context {
+	return metadata.AppendToOutgoingContext(ctx, "authorization", "Bearer "+*c.jwt.Load())
+}
+
+func (c *TaskCommandRouterClient) refreshJwt(ctx context.Context) error {
+	const jwtRefreshBufferSeconds = 30
+
+	if c.closed.Load() {
+		return errors.New("client is closed")
+	}
+
+	// If the current JWT expiration is already far enough in the future, don't refresh.
+	if exp := c.jwtExp.Load(); exp != nil && *exp-time.Now().Unix() > jwtRefreshBufferSeconds {
+		c.logger.DebugContext(ctx, "Skipping JWT refresh because expiration is far enough in the future", "task_id", c.taskID)
+		return nil
+	}
+
+	_, err, _ := c.refreshJwtGroup.Do("refresh", func() (any, error) {
+		if exp := c.jwtExp.Load(); exp != nil && *exp-time.Now().Unix() > jwtRefreshBufferSeconds {
+			return nil, nil
+		}
+
+		resp, err := c.serverClient.TaskGetCommandRouterAccess(ctx, pb.TaskGetCommandRouterAccessRequest_builder{
+			TaskId: c.taskID,
+		}.Build())
+		if err != nil {
+			return nil, fmt.Errorf("failed to refresh JWT: %w", err)
+		}
+
+		if resp.GetUrl() != c.serverURL {
+			return nil, errors.New("task router URL changed during session")
+		}
+
+		jwt := resp.GetJwt()
+		c.jwt.Store(&jwt)
+		jwtExp := parseJwtExpiration(ctx, jwt, c.logger)
+		c.jwtExp.Store(jwtExp)
+		return nil, nil
+	})
+	return err
+}
+
+func (c *TaskCommandRouterClient) callWithAuthRetry(ctx context.Context, fn func(context.Context) error) error {
+	err := fn(c.authContext(ctx))
+	if err != nil {
+		if st, ok := status.FromError(err); ok && st.Code() == codes.Unauthenticated {
+			if refreshErr := c.refreshJwt(ctx); refreshErr != nil {
+				return refreshErr
+			}
+			return fn(c.authContext(ctx))
+		}
+	}
+	return err
+}
+
+// ExecStart starts a command execution.
+func (c *TaskCommandRouterClient) ExecStart(ctx context.Context, request *pb.TaskExecStartRequest) (*pb.TaskExecStartResponse, error) {
+	var resp *pb.TaskExecStartResponse
+	_, err := callWithRetriesOnTransientErrors(ctx, func() (struct{}, error) {
+		callErr := c.callWithAuthRetry(ctx, func(authCtx context.Context) error {
+			var err error
+			resp, err = c.stub.TaskExecStart(authCtx, request)
+			return err
+		})
+		return struct{}{}, callErr
+	}, defaultRetryOptions())
+	return resp, err
```

Comment:
This feels a little hard to follow because of how `resp` is mutated in the inner scope and then returned in the outer scope.

I opened a [PR targeting](https://github.com/modal-labs/libmodal/pull/244) your branch to showcase the approach. This way `ExecStart` can be written as:

```go
// ExecStart starts a command execution.
func (c *TaskCommandRouterClient) ExecStart(ctx context.Context, request *pb.TaskExecStartRequest) (*pb.TaskExecStartResponse, error) {
	return callWithRetriesOnTransientErrors(ctx, func() (*pb.TaskExecStartResponse, error) {
		return callWithAuthRetry(ctx, c, func(authCtx context.Context) (*pb.TaskExecStartResponse, error) {
			return c.stub.TaskExecStart(authCtx, request)
		})
	}, defaultRetryOptions())
}

```